### PR TITLE
KAFKA-15527: Support ResultOrder to reverseRange and reverseAll query over kv-store in IQv2 

### PR DESCRIPTION
**Pull request by hanyuzheng7**
Support `ResultOrder` to `reverseRange` and `reverseAll` query over kv-store in IQv2..

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: docs/streams/upgrade-guide.html [hunk 1]
```diff
@@ -136,7 +136,7 @@ <h3 class="anchor-heading"><a id="streams_notable_changes" class="anchor-link"><
     <h3><a id="streams_api_changes_370" href="#streams_api_changes_370">Streams API changes in 3.7.0</a></h3>
     <p>
         IQv2 supports <code>RangeQuery</code> that allows to specify unbounded, bounded, or half-open key-ranges, which return data in ascending (byte[]-lexicographical) order (per partition).
-        <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-985%3A+Add+reverseRange+and+reverseAll+query+over+kv-store+in+IQv2">KIP-985</a> extends this functionality by adding <code>.withDescendingKeys()<code> to allow user to receive data in descending order.
+        <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-985%3A+Add+reverseRange+and+reverseAll+query+over+kv-store+in+IQv2">KIP-985</a> extends this functionality by adding <code>.withDescendingKeys()</code> to allow user to receive data in descending order.
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html]: We should also add `.withAscendinKeys()`

File: streams/src/main/java/org/apache/kafka/streams/query/RangeQuery.java [hunk 2]
```diff
@@ -59,25 +58,25 @@ private RangeQuery(final Optional<K> lower, final Optional<K> upper, final boole
      * @param <V> The value type
      */
     public static <K, V> RangeQuery<K, V> withRange(final K lower, final K upper) {
-        return new RangeQuery<>(Optional.ofNullable(lower), Optional.ofNullable(upper), true);
+        return new RangeQuery<>(Optional.ofNullable(lower), Optional.ofNullable(upper), ResultOrder.ANY);
     }
 
     /**
-     * Determines if the serialized byte[] of the keys in ascending order.
+     * Determines if the serialized byte[] of the keys in ascending or any order.
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/query/RangeQuery.java]: Could also be descending.

File: streams/src/main/java/org/apache/kafka/streams/query/RangeQuery.java [hunk 3]
```diff
@@ -59,25 +58,25 @@ private RangeQuery(final Optional<K> lower, final Optional<K> upper, final boole
...
      * Order is based on the serialized byte[] of the keys, not the 'logical' key order.
-     * @return true if ascending, false otherwise.
+     * @return return the order of returned records based on the serialized byte[] of the keys(can be unordered, or in ascending or in descending order).
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/query/RangeQuery.java]: Missing space `keys(can` -> `keys (can`

File: streams/src/main/java/org/apache/kafka/streams/query/RangeQuery.java [hunk 4]
```diff
@@ -59,25 +58,25 @@ private RangeQuery(final Optional<K> lower, final Optional<K> upper, final boole
...
-        return isKeyAscending;
+    public ResultOrder resultOrder() {
+        return order;
     }
 
     /**
      * Set the query to return the serialized byte[] of the keys in descending order.
      * Order is based on the serialized byte[] of the keys, not the 'logical' key order.
-     * @return a new RangeQuery instance with descending flag set.
+     * @return a new RangeQuery instance with DESCENDING set.
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/query/RangeQuery.java]: I think the original description was better -- it's am implementation detail that we don't need to highlight that we use an enum value to represent it.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java [hunk 5]
```diff
@@ -254,12 +255,12 @@ private <R> QueryResult<R> runRangeQuery(final Query<R> query,
...
         final RangeQuery<K, V> typedQuery = (RangeQuery<K, V>) query;
         RangeQuery<Bytes, byte[]> rawRangeQuery;
-        final boolean isKeyAscending = typedQuery.isKeyAscending();
+        final ResultOrder order = typedQuery.resultOrder();
         rawRangeQuery = RangeQuery.withRange(
                 keyBytes(typedQuery.getLowerBound().orElse(null)),
                 keyBytes(typedQuery.getUpperBound().orElse(null))
         );
-        if (!isKeyAscending) {
+        if (order.equals(ResultOrder.DESCENDING)) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java:263]: Given that we need to add `withAscendingKeys()`, we also need to add this case here.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredTimestampedKeyValueStore.java [hunk 6]
```diff
@@ -268,12 +269,12 @@ private  <R> QueryResult<R> runRangeQuery(final Query<R> query,
...
         final RangeQuery<K, V> typedQuery = (RangeQuery<K, V>) query;
         RangeQuery<Bytes, byte[]> rawRangeQuery;
-        final boolean isKeyAscending = typedQuery.isKeyAscending();
+        ResultOrder order = typedQuery.resultOrder();
         rawRangeQuery = RangeQuery.withRange(
                 keyBytes(typedQuery.getLowerBound().orElse(null)),
                 keyBytes(typedQuery.getUpperBound().orElse(null))
         );
-        if (!isKeyAscending) {
+        if (order.equals(ResultOrder.DESCENDING)) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredTimestampedKeyValueStore.java:277]: as above: add ascending case

**Comment by mjsax** [comment]
There is also a checkstyle error on the build, that needs to be fixed.

File: docs/streams/upgrade-guide.html (hunk 1 above)
Comment (hanyuzheng7) [reply docs/streams/upgrade-guide.html]: ok

File: streams/src/main/java/org/apache/kafka/streams/query/RangeQuery.java (hunk 3 above)
Comment (hanyuzheng7) [reply streams/src/main/java/org/apache/kafka/streams/query/RangeQuery.java]: ok

File: streams/src/main/java/org/apache/kafka/streams/query/RangeQuery.java (hunk 4 above)
Comment (hanyuzheng7) [reply streams/src/main/java/org/apache/kafka/streams/query/RangeQuery.java]: ok

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java (hunk 5 above)
Comment (hanyuzheng7) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java:263]: ok

**Pull request merged**